Quotient/keyverificationsession.cpp

```diff
@@ -421,6 +509,70 @@ void KeyVerificationSession::handleMac(const KeyVerificationMacEvent& event)
 void KeyVerificationSession::trustKeys()
 {
     m_connection->database()->setSessionVerified(m_pendingEdKeyId);
+    m_connection->database()->setMasterKeyVerified(m_pendingMasterKey);
+    if (m_remoteUserId == m_connection->userId()) {
+        m_connection->reloadDevices();
+    }
+
+    if (m_pendingMasterKey.length() > 0) {
+        if (m_remoteUserId == m_connection->userId()) {
+            const auto selfSigningKey = m_connection->database()->loadEncrypted("m.cross_signing.self_signing"_ls);
+            if (!selfSigningKey.isEmpty()) {
+                QHash<QString, QHash<QString, QJsonObject>> signatures;
+                auto json = QJsonObject {
+                    {"keys"_ls, QJsonObject {
+                        {"ed25519:"_ls + m_remoteDeviceId, m_connection->edKeyForUserDevice(m_remoteUserId, m_remoteDeviceId)},
+                        {"curve25519:"_ls + m_remoteDeviceId, m_connection->curveKeyForUserDevice(m_remoteUserId, m_remoteDeviceId)},
+                    }},
+                    {"algorithms"_ls, QJsonArray {"m.olm.v1.curve25519-aes-sha2"_ls, "m.megolm.v1.aes-sha2"_ls}},
+                    {"device_id"_ls, m_remoteDeviceId},
+                    {"user_id"_ls, m_remoteUserId},
+                };
+                auto signature = sign(selfSigningKey, QJsonDocument(json).toJson(QJsonDocument::Compact));
+                json["signatures"_ls] = QJsonObject {
+                    {m_connection->userId(), QJsonObject {
+                        {"ed25519:"_ls + m_connection->database()->selfSigningPublicKey(), QString::fromLatin1(signature)},
+                    }},
+                };
+                signatures[m_remoteUserId][m_remoteDeviceId] = json;
+                auto uploadSignatureJob = m_connection->callApi<UploadCrossSigningSignaturesJob>(signatures);
+                connect(uploadSignatureJob, &BaseJob::finished, m_connection /*FIXME*/, [=](){
+                    //TODO error handling
+                });
+            } else {
+                // Not parenting to this since the session is going to be destroyed soon
+                // TODO: Delete the handler once it's finished
+                auto handler = new SSSSHandler(m_connection);
+                handler->setConnection(m_connection);
+                handler->unlockSSSSFromCrossSigning();
```

Comment:
Does it make sense to `deleteLater()` upon successful unlocking of the session? I realise it's not trivial atm because of several network requests flying - I'm working on something that might help that :)

Reply:
I'm considering to just *not* unlock SSSS automatically. that would clear up this problem a bit and would give the client more explicit control (and information of) the unlocking operation